### PR TITLE
[Merged by Bors] - refactor(Topology/Category): add CompHausLike.EffectiveEpi

### PR DESCRIPTION
**Pull request by dagurtomas**
This is the first part of the refactor of CompHaus and friends (adding the new API which will eventually be used to redefine the categories of compact Hausdorff spaces, see #12930)

---
- [x] depends on: #13904 
- [x] depends on: #13905 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Effective epimorphisms in `CompHausLike`
+
+In any category of compact Hausdorff spaces, continuous surjections are effective epimorphisms.
+-/
+
+universe u
+
+open CategoryTheory Limits
+
+attribute [local instance] CategoryTheory.ConcreteCategory.instFunLike
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean:19]:
```suggestion
attribute [local instance] ConcreteCategory.instFunLike
```

File: Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean:2]:
```suggestion
Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
```

File: Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace CompHausLike
+
+variable {P : TopCat.{u} → Prop}
+
+/--
+If `π` is a surjective morphism in `CompHausLike P`, then it is an effective epi.
+-/
+noncomputable
+def struct {B X : CompHausLike P} (π : X ⟶ B) (hπ : Function.Surjective π) :
```
Comment (jcommelin) [inline Mathlib/Topology/Category/CompHausLike/EffectiveEpi.lean:29]:
```suggestion
def effectiveEpiStruct {B X : CompHausLike P} (π : X ⟶ B) (hπ : Function.Surjective π) :
```

**Review by jcommelin — COMMENTED** [review]
1 minor comment

bors d+

**Comment by dagurtomas** [comment]
Thanks!

bors merge

**Pull request closed without merging**